_posts/2024-07-01-zio-fullstack-webapp.md

```diff
@@ -0,0 +1,844 @@
+
+
+## 6. Tapir
+
+Tapir is a library for defining API endpoint as immutable data structures.
+
+This is a huge win, because it allows to:
+
+* define the API once
+* generate the client and server side
+* generate the documentation
+* generate the tests
+
+Here is an example of Tapir endpoints definition:
+
+### 6.1. Shared Endpoint definitions
+
+In the shared module, the API endpoints are defined as immutable data structures.
+
+```scala
+object PersonEndpoints {
+  val get Endpoint[Unit, String, Person, Any] = endpoint.get
+    .in("person")
+    .out(jsonBody[Person])
+    .errorOut(stringBody)
+    .summary("Get a person by id")
+
+  val create: Endpoint[Unit, Person, Throwable, User, Any] = baseEndpoint
+    .tag("person")
+    .name("person")
+    .post
+    .in("person")
+    .in(
+      jsonBody[Person]
+        .description("Person to create")
+        .example(Person("John", 30, Left(Cat("Fluffy"))))
+    )
+    .out(jsonBody[User])
+    .description("Create person")
+
+}
+```
+
+This is a simple GET endpoint that will return a Person and a POST endpoint that will create a Person and return a User.
+
+From this definition, Tapir will help to generate the client and server side code, the documentation and the tests.
+
+
+### 6.2. Server side
+
+On **server side**, Tapir Endpoints.create will be implemented by a ZIO effect, through the **zServerLogic** combinator.
+
+```scala
+                                           // (1)
+class PersonController private (personService: PersonService) extends BaseController {
+  [...]
+
+  val create: ServerEndpoint[Any, Task] = PersonEndpoint.createEndpoint
+    .zServerLogic( // (2)
+      p => personService.register(p)
+    )
+
+  val routes: List[ServerEndpoint[Any, Task]] =
+    List(create)
+}
+
+object PersonController {
+  def makeZIO: URIO[PersonService, PersonController] = // (3)
+    ZIO.service[PersonService].map(new PersonController(_))
+}
+
+```
+
+* **(1)** Dependency Injection.
+* **(2)** Controller implementation, where json codec are transparently handled.
+* **(3)** Service instantiation:  URIO[PersonService, PersonController]
+
+
+Everything is IO, URIO[PersonService, PersonController] is an IO that needs a PersonService and will produce a PersonController and cannot fail.
+
+> **URIO** is a type alias for a ZIO that cannot fail:
+>
+>   type **URIO**[R, A] = ZIO[R, **Nothing**, A]`
+
+**zServerLogic** is part of the Tapir ZIO support, and bridges a ZIO in the Rest API.
+
+**zServerLogic** is a combinator that will transform a ZIO effect into a server logic, it will handle the error and the success case.
```

Comment:
Can you also describe some alternatives? For example, in the ZIO Rite of Passage course I use `serverLogic` and `serverLogicSuccess`. No need to list them all, just make the reader aware that there are alternatives